c_binding/libsg.h

```diff
@@ -0,0 +1,141 @@
+ */
+
+#ifndef _LIBSG_H_
+#define _LIBSG_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+
+#include "libstoragemgmt/libstoragemgmt_common.h"
+
+#define _SG_T10_SPC_ASSOCIATION_TGT_PORT            1
+/* ^ SPC-5 rev 7, Table 487 - ASSOCIATION field */
+
+#define _SG_T10_SPC_VPD_DI_DESIGNATOR_TYPE_NAA      0x3
+
+#define _SG_T10_SPL_SAS_ADDR_LEN                    17
+/* ^ SPL-4 rev5 4.2.4 SAS address. With trailing \0. */
+#define _SG_T10_SPL_SAS_ADDR_LEN_BITS               8
+/* ^ SPL-4 rev5 4.2.4 SAS address. */
+
+#define _SG_T10_SPC_VPD_DI                          0x83
+#define _SG_T10_SPC_VPD_DI_NAA_235_ID_LEN           8
+#define _SG_T10_SPC_VPD_DI_NAA_6_ID_LEN             16
+#define _SG_T10_SPC_VPD_DI_NAA_TYPE_2               0x2
+#define _SG_T10_SPC_VPD_DI_NAA_TYPE_3               0x3
+#define _SG_T10_SPC_VPD_DI_NAA_TYPE_5               0x5
+#define _SG_T10_SPC_VPD_DI_NAA_TYPE_6               0x6
+#define _SG_T10_SPC_VPD_DI_ASSOCIATION_LUN          0
+
+#define _SG_T10_SPC_VPD_SUP_VPD_PGS                 0x00
+/* ^ SPC-5 rev 7, 7.7.16 Supported VPD Pages VPD page */
+
+#define _SG_T10_SPC_INQUERY_MAX_LEN                 0xffff
+/* ^ SPC-5 rev 7, Table 142 - INQUIRY command */
+#define _SG_T10_SPC_VPD_MAX_LEN                     _SG_T10_SPC_INQUERY_MAX_LEN
+/* ^ VPD is a INQUERY */
+#define _SG_T10_SPC_RECV_DIAG_MAX_LEN               0xffff
+/* ^ SPC-5 rev 7, Table 219 - RECEIVE DIAGNOSTIC RESULTS command */
+#define _SG_T10_SPC_SEND_DIAG_MAX_LEN               0xffff
+/* ^ SPC-5 rev 7, Table 269 - SEND DIAGNOSTIC command */
+
+#define _SG_T10_SPC_PROTOCOL_ID_SAS                 6
+/* ^ SPC-5 rev 7, Table 444 - PROTOCOL IDENTIFIER field values */
+
+
+#pragma pack(push, 1)
+/*
+ * Table 589 - Device Identification VPD page
+ */
+LSM_DLL_LOCAL struct _sg_t10_vpd83_header {
+    uint8_t dev_type : 5;
+    /* ^ PERIPHERAL DEVICE TYPE */
+    uint8_t qualifier : 3;
+    /* PERIPHERAL QUALIFIER */
+    uint8_t page_code;
+    uint16_t page_len;
+};
+
+/*
+ * Table 590 - Designation descriptor
+ */
+LSM_DLL_LOCAL struct _sg_t10_vpd83_dp_header {
+    uint8_t code_set        : 4;
+    uint8_t protocol_id     : 4;
+    uint8_t designator_type : 4;
+    uint8_t association     : 2;
+    uint8_t reserved_1      : 1;
+    uint8_t piv             : 1;
+    uint8_t reserved_2;
+    uint8_t designator_len;
+};
+
+LSM_DLL_LOCAL struct _sg_t10_vpd83_dp {
+    struct _sg_t10_vpd83_dp_header header;
+    uint8_t designator[0xff];
+};
+/* ^ SPC-5 rev 7, Table 486 - Designation descriptor. */
+
+LSM_DLL_LOCAL struct _sg_t10_vpd83_naa_header {
+    uint8_t data_msb : 4;
+    uint8_t naa_type : 4;
+};
+
+#pragma pack(pop)
+
+/*
+ * Caller should make sure fd is not NULL.
+ */
+LSM_DLL_LOCAL int _sg_io_open_ro(char *err_msg, const char *disk_path, int *fd);
+
+LSM_DLL_LOCAL int _sg_io_open_rw(char *err_msg, const char *disk_path, int *fd);
+
+/*
+ * Assuming err_msg and vpd_data are not NULL, caller should make sure that.
+ */
```

Comment:
No. That's a good suggestion.
Even those are internal functions, I think I should enforce that by `assert()` just in case other dev missed the function comments.